mirgecom/steppers.py

```diff
@@ -59,6 +58,17 @@ def advance_state(rhs, timestepper, checkpoint, get_timestep,
         Time at which to start
     istep: int
         Step number from which to start
+    pre_step_callback
+        An optional user-defined function to be called before the timestepper
+        is called for that particular step. A non-zero return code from this
+        function indicates that this function should stop gracefully.
+        Examples of such functions include visualization, io, or restart
+        checkpoints.
```

Comment:
Ultimately I think checkpointing will be better off inside `post_step_callback`, but switching that up is probably beyond the scope of this PR.

Reply:
@majosm I actually think that shouldn't be hard at all; it would just mean you pass your checkpoint function to the `post_step_callback` arg. Or am I missing something?

Reply:
> @majosm I actually think that shouldn't be hard at all; it would just mean you pass your checkpoint function to the `post_step_callback` arg. Or am I missing something?

IIRC there were some complications with making sure that you get one and only one checkpoint call per step (including step 0), especially when restarts are involved. I don't remember if this was fully resolved or not in #257.

Reply:
If you have something separate in-place to make your time0 dumps, then having checkpoint as post-step callback is OK, but I don't see the advantage or the need.   Checkpoint is currently pre-step because users often want to dump viz files before stepping, and simulation health check gets a change to stop the simulation before stepping occurs with a bad state.

Reply:
> If you have something separate in-place to make your time0 dumps, then having checkpoint as post-step callback is OK, but I don't see the advantage or the need. Checkpoint is currently pre-step because users often want to dump viz files before stepping, and simulation health check gets a change to stop the simulation before stepping occurs with a bad state.

IIRC doing it before the step led to duplicated checkpoints on restart, e.g.:

```
....
----- start step N -----
checkpoint
advance
----- end step N -------
final checkpoint
===== end run 1 ========
===== start run 2 ======
----- start step N+1 ---
checkpoint  # oops, same as final checkpoint from run 1
advance
----- end step N+1 -----
```

Easier to handle this with post-step checkpoints + a step 0 checkpoint IMO.

Reply:
> (...)
> IIRC doing it before the step led to duplicated checkpoints on restart, e.g.:
> (...)
> Easier to handle this with post-step checkpoints + a step 0 checkpoint IMO.

Thumbs up, but also further motivation for getting viz and healthcheck detangled from simutils checkpoint (not that healthcheck is tangled [yet!]).  I think simutils checkpoint could easily just be rolled into a `fluids_checkpoint` and then call it from the user's `my_checkpoint`, then we could have a similar thing for each (non-fluid) operator too.  I think we were discussing something along those lines in #257 before.

Reply:
> Thumbs up, but also further motivation for getting viz and healthcheck detangled from simutils checkpoint

I believe this accomplishes that with minimal pain